Core/clim-basic/extended-streams/stream-input.lisp

```diff
@@ -365,23 +365,25 @@ keys read."))
 
 (defmethod stream-read-gesture ((stream standard-extended-input-stream)
                                 &key &allow-other-keys)
-  (multiple-value-bind (gesture unavailable-reason)
-      (call-next-method)
-    (if (null gesture)
-        (values nil unavailable-reason)
-        (flet ((abort-gesture-p (gesture)
-                 (loop for gesture-name in *abort-gestures*
-                         thereis (event-matches-gesture-name-p gesture gesture-name)))
-               (accelerator-gesture-p (gesture)
-                 (loop for gesture-name in *accelerator-gestures*
-                         thereis (event-matches-gesture-name-p gesture gesture-name))))
-          (cond
-            ((abort-gesture-p gesture)
-             (signal 'abort-gesture :event gesture))
-            ((accelerator-gesture-p gesture)
-             (signal 'accelerator-gesture :event gesture))
-            (t
-             (return-from stream-read-gesture gesture)))))))
+  (loop
+    (multiple-value-bind (gesture unavailable-reason)
+        (call-next-method)
+      (if (null gesture)
+          (return-from stream-read-gesture
+            (values nil unavailable-reason))
+          (flet ((abort-gesture-p (gesture)
+                   (loop for gesture-name in *abort-gestures*
+                           thereis (event-matches-gesture-name-p gesture gesture-name)))
+                 (accelerator-gesture-p (gesture)
+                   (loop for gesture-name in *accelerator-gestures*
+                           thereis (event-matches-gesture-name-p gesture gesture-name))))
+            (cond
+              ((abort-gesture-p gesture)
+               (signal 'abort-gesture :event gesture))
+              ((accelerator-gesture-p gesture)
+               (signal 'accelerator-gesture :event gesture))
+              (t
+               (return-from stream-read-gesture gesture))))))))
```

Comment:
How do we know that we must loop instead of returning the gesture?

Reply:
This is not explicitly specified. the thing is that what we did before was wrong (because stream-read-gesture returned nil when the condition was not handled).

We must either loop or return the gesture. Looping has a wonderful benefit though -  you may change the command table (and the set of acceleration/abort gestures) dynamically, or simly duly note the value of the accelerator and continue processing where it was.

That will allow baking into our command processing gesture combos (without compromising backward compatibility nor requireing custom command processors like ESA).